File: content_base/bundle/needleman_wunsch.py
from typing import List
from functools import reduce
from pprint import pprint, pformat


class Direction:
    def __init__(self, val: int = 0b000):
        """
        self.val is a representation of direction
        up:      0b001
        left:    0b010
        upleft:  0b100
        default: 0b000

        each direction acts as a flag, thus it can be overrapped
        """
        self.val = val

    def __repr__(self):
        rep_list = []
        if self.val & 0b001:  # test it contains up direction
            rep_list.append("↑")
        if self.val & 0b010:  # test it contains left direction
            rep_list.append("←")
        if self.val & 0b100:  # test it contains upleft direction
            rep_list.append("↖")

        if len(rep_list) == 0:
            return "-"
        return "|".join(rep_list)


def needleman_wunsch(x: str, y: str) -> str:
    def calc_element_value(
        row: int,
        col: int,
        /,
        match_value: int = 1,
        mis_match_value: int = -1,
        gap_penalty: int = -1,
    ) -> tuple:
        """
        closure for calculate matrix with free variable:
        matrix, tx, ty
        """
        assert row > 0, f"invalid row index({row}) found"
        assert col > 0, f"invalid col index({col}) found"

        diag_element = matrix[row - 1][col - 1]
        left_element = matrix[row][col - 1]
        up_element = matrix[row - 1][col]

        diag_value, _diag_direction = diag_element
        l_value, _l_direction = left_element
        up_value, _up_direction = up_element

        s_value = mis_match_value
        if tx[row] == ty[col]:
            s_value = match_value

        candidate_l = l_value + gap_penalty
        candidate_up = up_value + gap_penalty
        candidate_diag = diag_value + s_value

        element_value = max(candidate_l, candidate_up, candidate_diag)
        element_direction = Direction()

        if element_value == candidate_up:
            element_direction.val |= 0b001
        if element_value == candidate_l:
            element_direction.val |= 0b010
        if element_value == candidate_diag:
            element_direction.val |= 0b100

        matrix[row][col] = (element_value, element_direction)

        return matrix[row][col]

    if len(x) == 0 or len(y) == 0:
        return ""

    tx = f" {x}"
    ty = f" {y}"
    row_cnt = len(tx)
    col_cnt = len(ty)

    matrix = [[(0, Direction())] * col_cnt for _ in range(row_cnt)]
    # initialize first row
    for idx, _ in enumerate(matrix[0]):
        matrix[0][idx] = (-idx, Direction())
    # initialize first column
    for idx, _ in enumerate(matrix):
        matrix[idx][0] = (-idx, Direction())

    for ix in range(1, row_cnt):
        for iy in range(1, col_cnt):
            calc_element_value(ix, iy)
    visualized_matrix(tx, ty, matrix)

    cur_row = row_cnt - 1
    cur_col = col_cnt - 1
    x_result = ""
    y_result = ""

    stack = [(cur_row, cur_col, x_result, y_result)]

    result_list = []
    result_set = set()

    while stack:
        cur_row, cur_col, x_result, y_result = stack.pop()
        # base condition
        if cur_row == 0:
            y_result = cur_col * "-" + y_result
            if (x_result, y_result) not in result_set:
                result_set.add((x_result, y_result))
                result_list.append((x_result, y_result))
            if len(stack) > 1:
                stack.pop()
            continue
        if cur_col == 0:
            x_result = cur_row * "-" + x_result
            if (x_result, y_result) not in result_set:
                result_set.add((x_result, y_result))
                result_list.append((x_result, y_result))
            if len(stack) > 1:
                stack.pop()
            continue

        cur_item = matrix[cur_row][cur_col]
        cur_value, cur_direction = cur_item

        if cur_direction.val & 0b001:
            stack.append((cur_row - 1, cur_col, "-" + x_result, y_result))
        if cur_direction.val & 0b010:
            stack.append((cur_row, cur_col - 1, x_result, "-" + y_result))
        if cur_direction.val & 0b100:
            stack.append(
                (
                    cur_row - 1,
                    cur_col - 1,
                    tx[cur_row] + x_result,
                    ty[cur_col] + y_result,
                )
            )
        if cur_direction.val == 0b000:
            print("error")
            break
    return result_list


def visualized_matrix(tx: str, ty: str, matrix: List):
    print(f"•|" + "          |".join(ty))
    for idx, _ in enumerate(matrix):
        row_repr = reduce(reducer, matrix[idx], tx[idx])
        print(row_repr)


def reducer(acc, cur) -> str:
    value, direction = cur
    return f"{acc}|({value:+03},{repr(direction):>5})"


if __name__ == "__main__":
    x = "XMJYAUZ"
    y = "MZJAWXU"
    result = needleman_wunsch(x, y)
    print(result)

File: content_base/bundle/NeedlemanWunsch.kt
class Direction(
    var value: Int,
) {
    override fun toString(): String {
        if (this.value == 0b000) return "-"
        val reprList: MutableList<String> = mutableListOf()
        if (this.value and 0b001 != 0) reprList.add("↑")
        if (this.value and 0b010 != 0) reprList.add("←")
        if (this.value and 0b100 != 0) reprList.add("↖")
        return reprList.joinToString("|")
    }
}

data class Tuple(val row: Int, val col: Int, val xRes: String, val yRes: String)

fun needlemanWunsch(
    x: String,
    y: String,
): String {
    val matrix: Array<Array<Pair<Int, Direction>>> = Array(x.length + 1) { Array(y.length + 1) { 0 to Direction(0b000) } }
    val tx: String = " $x"
    val ty: String = " $y"

    fun calcElementValue(
        row: Int,
        col: Int,
        matchValue: Int = 1,
        misMatchValue: Int = -1,
        gapPanalty: Int = -1,
    ): Pair<Int, Direction> {
        // closure for calculate matrix with free variables:
        // matrix, tx, ty
        require(row > 0) { "Invalid row index($row) found" }
        require(col > 0) { "Invalid column index($col) found" }

        val (upValue, upDirection) = matrix[row - 1][col]
        val (leftValue, leftDirection) = matrix[row][col - 1]
        val (diagValue, diagDirection) = matrix[row - 1][col - 1]

        val upCandidate = upValue + gapPanalty
        val leftCandidate = leftValue + gapPanalty

        var diagCandidate = diagValue + misMatchValue
        if (tx[row] == ty[col]) {
            diagCandidate = diagValue + matchValue
        }

        var elementDirection: Direction = Direction(0b000)

        val elementValue: Int = maxOf(upCandidate, leftCandidate, diagCandidate)
        if (upCandidate == elementValue) {
            elementDirection.value = elementDirection.value or 0b001
        }
        if (leftCandidate == elementValue) {
            elementDirection.value = elementDirection.value or 0b010
        }
        if (diagCandidate == elementValue) {
            elementDirection.value = elementDirection.value or 0b100
        }

        matrix[row][col] = elementValue to elementDirection
        return matrix[row][col]
    }

    // initialize first column
    for (i in matrix.indices) {
        matrix[i][0] = -i to Direction(0b000)
    }
    // initialize first row
    for (j in matrix[0].indices) {
        matrix[0][j] = -j to Direction(0b000)
    }

    for (i in matrix.indices) {
        for (j in matrix[i].indices) {
            if (i == 0 || j == 0) continue
            calcElementValue(i, j)
        }
    }

    visualizeMatrix(tx, ty, matrix)

    var curRow = x.length
    var curCol = y.length
    var xResult = ""
    var yResult = ""

    var stack = ArrayDeque<Tuple>()
    stack.add(Tuple(curRow, curCol, xResult, yResult))

    val resultList: MutableList<Pair<String, String>> = mutableListOf()

    while (stack.size > 0) {
        val curEle = stack.removeLast()
        curRow = curEle.row
        curCol = curEle.col
        xResult = curEle.xRes
        yResult = curEle.yRes

        if (curRow == 0) {
            yResult = "-".repeat(curCol) + yResult
            resultList.add(xResult to yResult)
            if (stack.size > 0) {
                stack.removeLast()
            }
            continue
        }

        if (curCol == 0) {
            xResult = "-" .repeat(curRow) + xResult
            resultList.add(xResult to yResult)
            if (stack.size > 0) {
                stack.removeLast()
            }
            continue
        }

        val (curValue, curDirection) = matrix[curRow][curCol]
        if (curDirection.value and 0b001 != 0) {
            stack.add(Tuple(curRow - 1, curCol, "-" + xResult, yResult))
        }
        if (curDirection.value and 0b010 != 0) {
            stack.add(Tuple(curRow, curCol - 1, xResult, "-" + yResult))
        }
        if (curDirection.value and 0b100 != 0) {
            xResult = tx[curRow] + xResult
            yResult = ty[curCol] + yResult
            stack.add(Tuple(curRow - 1, curCol - 1, xResult, yResult))
        }
    }

    return resultList.joinToString("|")
}

fun visualizeMatrix(
    tx: String,
    ty: String,
    matrix: Array<Array<Pair<Int, Direction>>>,
) {
    println("•|${ty.toList().joinToString("          |")}")
    for (idx in matrix.indices) {
        val rowRepr = matrix[idx].joinToString("|") { "(${ "%+03d".format(it.first) },${ "%5s".format(it.second) })" }
        println("${tx[idx]}|$rowRepr")
    }
}

fun main() {

    // val x = "XMJYAUZ"
    // val y = "MZJAWXU"
    val x = "Reading out a LCS"
    val y = "Reading out all LCSs"
    val res = needlemanWunsch(x, y)
    println(res)
}
